File: wave.rb
# wave.rb
require 'optparse'
require 'fileutils'

class Envelope
  attr_reader :attack, :decay, :sustain, :release
  def initialize(a, d, s, r)
    @attack, @decay, @sustain, @release = a, d, s, r
  end
end

class WaveGenerator
  attr_accessor :freq, :amp, :duration, :sample_rate, :wave_type, :stereo, :envelope, :sweep_start, :sweep_end

  def initialize(opts = {})
    @freq = opts[:freq] || 440
    @amp = opts[:amp] || 0.8
    @duration = opts[:duration] || 2.0
    @sample_rate = opts[:sample_rate] || 44100
    @wave_type = opts[:wave_type] || 'sine'
    @stereo = opts[:stereo] || false
    @envelope = opts[:envelope]
    @sweep_start = opts[:sweep_start]
    @sweep_end = opts[:sweep_end]
    @samples = []
  end

  def value_at(t)
    freq = @freq
    if @sweep_start && @sweep_end
      freq = @sweep_start + (@sweep_end - @sweep_start) * (t / @duration)
    end
    phase = 2 * Math::PI * freq * t
    val = case @wave_type
          when 'sine' then Math.sin(phase)
          when 'square' then Math.sin(phase) >= 0 ? 1 : -1
          when 'sawtooth' then 2 * (phase / (2 * Math::PI) - (phase / (2 * Math::PI) + 0.5).floor)
          when 'triangle' then 2 * (2 * (phase / (2 * Math::PI) - (phase / (2 * Math::PI) + 0.5).floor).abs) - 1
          when 'noise' then rand(-1.0..1.0)
          else 0
          end
    val *= envelope_value(t) if @envelope
    val
  end

  def envelope_value(t)
    e = @envelope
    if t < e.attack
      return t / (e.attack > 0 ? e.attack : 1)
    end
    t -= e.attack
    if t < e.decay
      return 1 - (1 - e.sustain) * (t / (e.decay > 0 ? e.decay : 1))
    end
    t -= e.decay
    sustain_dur = @duration - e.attack - e.decay - e.release
    if sustain_dur > 0 && t < sustain_dur
      return e.sustain
    end
    t -= sustain_dur
    if t < e.release
      return e.sustain * (1 - t / (e.release > 0 ? e.release : 1))
    end
    0
  end

  def generate
    num_samples = (@duration * @sample_rate).to_i
    channels = @stereo ? 2 : 1
    @samples = Array.new(channels) { [] }
    num_samples.times do |i|
      t = i / @sample_rate.to_f
      val = value_at(t)
      sample = (@amp * val * 32767).round
      channels.times { |ch| @samples[ch] << sample }
    end
  end

  def save_wav(filename)
    File.open(filename, 'wb') do |f|
      channels = @samples.size
      num_samples = @samples[0].size
      data_size = num_samples * 2 * channels
      byte_rate = @sample_rate * 2 * channels
      f.write('RIFF')
      f.write([36 + data_size].pack('V'))
      f.write('WAVE')
      f.write('fmt ')
      f.write([16, 1, channels, @sample_rate, byte_rate, 2 * channels, 16].pack('VvvVVvv'))
      f.write('data')
      f.write([data_size].pack('V'))
      num_samples.times do |i|
        channels.times { |ch| f.write([@samples[ch][i]].pack('v')) }
      end
    end
  end

  def play(filename)
    os = RUBY_PLATFORM
    cmd = nil
    if os =~ /mswin|mingw|windows/
      cmd = "start #{filename}"
    elsif os =~ /darwin/
      cmd = "afplay #{filename}"
    else
      players = ['ffplay', 'aplay', 'mpg123']
      found = players.find { |p| system("which #{p} > /dev/null 2>&1") }
      cmd = "#{found} #{filename}" if found
    end
    system(cmd) if cmd
  end
end

def interactive
  puts "=== Waveform Audio Generator ==="
  types = ['sine', 'square', 'sawtooth', 'triangle', 'noise']
  print "Type (#{types.join('/')}) [sine]: "
  wtype = gets.chomp.strip
  wtype = 'sine' if wtype.empty? || !types.include?(wtype)
  print "Frequency (Hz) [440]: "
  freq = gets.to_f; freq = 440 if freq <= 0
  print "Amplitude (0.0-1.0) [0.8]: "
  amp = gets.to_f; amp = 0.8 if amp < 0 || amp > 1
  print "Duration (s) [2.0]: "
  dur = gets.to_f; dur = 2.0 if dur <= 0
  print "Sample rate [44100]: "
  rate = gets.to_i; rate = 44100 if rate <= 0
  print "Stereo? (y/n) [n]: "
  stereo = gets.chomp.downcase == 'y'
  print "ADSR? (Attack Decay Sustain Release) [none]: "
  adsr_str = gets.chomp.strip
  env = nil
  if adsr_str != '' && adsr_str != 'none'
    parts = adsr_str.split.map(&:to_f)
    env = Envelope.new(*parts) if parts.size == 4
  end
  print "Output file [output.wav]: "
  fname = gets.chomp.strip
  fname = 'output.wav' if fname.empty?
  fname += '.wav' unless fname.end_with?('.wav')
  gen = WaveGenerator.new(freq: freq, amp: amp, duration: dur, sample_rate: rate,
                          wave_type: wtype, stereo: stereo, envelope: env)
  gen.generate
  gen.save_wav(fname)
  puts "Saved to #{fname}"
  print "Play? (y/n) [y]: "
  play = gets.chomp.downcase
  gen.play(fname) unless play == 'n'
end

def cli
  options = {}
  OptionParser.new do |opts|
    opts.on('-f FREQ', '--freq FREQ', Float) { |v| options[:freq] = v }
    opts.on('-t TYPE', '--type TYPE') { |v| options[:type] = v }
    opts.on('-a AMP', '--amp AMP', Float) { |v| options[:amp] = v }
    opts.on('-d DUR', '--dur DUR', Float) { |v| options[:dur] = v }
    opts.on('-r RATE', '--rate RATE', Integer) { |v| options[:rate] = v }
    opts.on('-o OUTPUT', '--output OUTPUT') { |v| options[:output] = v }
    opts.on('--stereo', 'Stereo') { options[:stereo] = true }
    opts.on('--adsr ADSR', 'Attack Decay Sustain Release') { |v| options[:adsr] = v }
    opts.on('--sweep SWEEP', 'Start End') { |v| options[:sweep] = v }
    opts.on('--no-play', 'Skip playback') { options[:no_play] = true }
  end.parse!
  freq = options[:freq] || 440
  wtype = options[:type] || 'sine'
  amp = options[:amp] || 0.8
  dur = options[:dur] || 2.0
  rate = options[:rate] || 44100
  stereo = options[:stereo] || false
  output = options[:output] || 'output.wav'
  env = nil
  if options[:adsr]
    parts = options[:adsr].split.map(&:to_f)
    env = Envelope.new(*parts) if parts.size == 4
  end
  sweep_start = sweep_end = nil
  if options[:sweep]
    parts = options[:sweep].split.map(&:to_f)
    sweep_start, sweep_end = parts if parts.size == 2
  end
  gen = WaveGenerator.new(freq: freq, amp: amp, duration: dur, sample_rate: rate,
                          wave_type: wtype, stereo: stereo, envelope: env,
                          sweep_start: sweep_start, sweep_end: sweep_end)
  gen.generate
  gen.save_wav(output)
  puts "Saved to #{output}"
  gen.play(output) unless options[:no_play]
end

if ARGV.empty?
  interactive
else
  cli
end

File: wave.js
// wave.js
const fs = require('fs');
const readline = require('readline');
const { exec } = require('child_process');

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

class Envelope {
    constructor(attack, decay, sustain, release) {
        this.attack = attack;
        this.decay = decay;
        this.sustain = sustain;
        this.release = release;
    }
}

class WaveGenerator {
    constructor(opts = {}) {
        this.freq = opts.freq || 440;
        this.amp = opts.amp || 0.8;
        this.duration = opts.duration || 2.0;
        this.sampleRate = opts.sampleRate || 44100;
        this.waveType = opts.waveType || 'sine';
        this.stereo = opts.stereo || false;
        this.envelope = opts.envelope || null;
        this.sweepStart = opts.sweepStart || null;
        this.sweepEnd = opts.sweepEnd || null;
        this.samples = [];
    }

    _valueAt(t) {
        let freq = this.freq;
        if (this.sweepStart !== null && this.sweepEnd !== null) {
            freq = this.sweepStart + (this.sweepEnd - this.sweepStart) * (t / this.duration);
        }
        const phase = 2 * Math.PI * freq * t;
        let val = 0;
        switch (this.waveType) {
            case 'sine': val = Math.sin(phase); break;
            case 'square': val = Math.sin(phase) >= 0 ? 1 : -1; break;
            case 'sawtooth': val = 2 * (phase / (2 * Math.PI) - Math.floor(phase / (2 * Math.PI) + 0.5)); break;
            case 'triangle': val = 2 * Math.abs(2 * (phase / (2 * Math.PI) - Math.floor(phase / (2 * Math.PI) + 0.5))) - 1; break;
            case 'noise': val = Math.random() * 2 - 1; break;
            default: val = 0;
        }
        if (this.envelope) val *= this._envelopeValue(t);
        return val;
    }

    _envelopeValue(t) {
        const e = this.envelope;
        if (t < e.attack) return t / e.attack || 1;
        let t2 = t - e.attack;
        if (t2 < e.decay) return 1 - (1 - e.sustain) * (t2 / e.decay || 0);
        t2 -= e.decay;
        const sustainDur = this.duration - e.attack - e.decay - e.release;
        if (sustainDur > 0 && t2 < sustainDur) return e.sustain;
        t2 -= sustainDur;
        if (t2 < e.release) return e.sustain * (1 - t2 / e.release || 0);
        return 0;
    }

    generate() {
        const numSamples = Math.floor(this.duration * this.sampleRate);
        const channels = this.stereo ? 2 : 1;
        this.samples = Array.from({ length: channels }, () => []);
        for (let i = 0; i < numSamples; i++) {
            const t = i / this.sampleRate;
            const val = this._valueAt(t);
            const sample = Math.round(this.amp * val * 32767);
            for (let ch = 0; ch < channels; ch++) {
                this.samples[ch].push(sample);
            }
        }
    }

    saveWAV(filename) {
        const channels = this.samples.length;
        const numSamples = this.samples[0].length;
        const dataSize = numSamples * 2 * channels;
        const byteRate = this.sampleRate * 2 * channels;
        const buffer = Buffer.alloc(44 + dataSize);
        let offset = 0;
        const writeU32 = v => { buffer.writeUInt32LE(v, offset); offset += 4; };
        const writeU16 = v => { buffer.writeUInt16LE(v, offset); offset += 2; };
        buffer.write('RIFF', 0);
        writeU32(36 + dataSize);
        buffer.write('WAVE', 8);
        buffer.write('fmt ', 12);
        writeU32(16);
        writeU16(1);
        writeU16(channels);
        writeU32(this.sampleRate);
        writeU32(byteRate);
        writeU16(2 * channels);
        writeU16(16);
        buffer.write('data', offset);
        offset += 4;
        writeU32(dataSize);
        for (let i = 0; i < numSamples; i++) {
            for (let ch = 0; ch < channels; ch++) {
                buffer.writeInt16LE(this.samples[ch][i], offset);
                offset += 2;
            }
        }
        fs.writeFileSync(filename, buffer);
    }

    play(filename) {
        const platform = process.platform;
        let cmd;
        if (platform === 'win32') cmd = `start "" "${filename}"`;
        else if (platform === 'darwin') cmd = `afplay "${filename}"`;
        else cmd = `ffplay "${filename}" || aplay "${filename}" || mpg123 "${filename}"`;
        exec(cmd, (err) => { if (err) console.log('Playback error:', err.message); });
    }
}

function ask(q) { return new Promise(r => rl.question(q, r)); }

async function interactive() {
    console.log('=== Waveform Audio Generator ===');
    const types = ['sine', 'square', 'sawtooth', 'triangle', 'noise'];
    let wtype = await ask(`Type (${types.join('/')}) [sine]: `) || 'sine';
    if (!types.includes(wtype)) wtype = 'sine';
    const freq = parseFloat(await ask('Frequency (Hz) [440]: ') || '440') || 440;
    const amp = parseFloat(await ask('Amplitude (0.0-1.0) [0.8]: ') || '0.8') || 0.8;
    const dur = parseFloat(await ask('Duration (s) [2.0]: ') || '2.0') || 2.0;
    const rate = parseInt(await ask('Sample rate [44100]: ') || '44100') || 44100;
    const stereo = (await ask('Stereo? (y/n) [n]: ')).toLowerCase() === 'y';
    const adsrStr = await ask('ADSR? (Attack Decay Sustain Release) [none]: ');
    let env = null;
    if (adsrStr && adsrStr !== 'none') {
        const parts = adsrStr.split(/\s+/).map(Number);
        if (parts.length === 4) env = new Envelope(...parts);
    }
    let fname = await ask('Output file [output.wav]: ') || 'output.wav';
    if (!fname.endsWith('.wav')) fname += '.wav';
    const gen = new WaveGenerator({ freq, amp, duration: dur, sampleRate: rate, waveType: wtype, stereo, envelope: env });
    gen.generate();
    gen.saveWAV(fname);
    console.log(`Saved to ${fname}`);
    if ((await ask('Play? (y/n) [y]: ')).toLowerCase() !== 'n') gen.play(fname);
    rl.close();
}

function cli() {
    const args = require('minimist')(process.argv.slice(2));
    let env = null;
    if (args.adsr) {
        const parts = args.adsr.split(/\s+/).map(Number);
        if (parts.length === 4) env = new Envelope(...parts);
    }
    let sweepStart = null, sweepEnd = null;
    if (args.sweep) {
        const parts = args.sweep.split(/\s+/).map(Number);
        if (parts.length === 2) { sweepStart = parts[0]; sweepEnd = parts[1]; }
    }
    const gen = new WaveGenerator({
        freq: args.f || args.freq || 440,
        amp: args.a || args.amp || 0.8,
        duration: args.d || args.dur || 2.0,
        sampleRate: args.r || args.rate || 44100,
        waveType: args.t || args.type || 'sine',
        stereo: args.stereo || false,
        envelope: env,
        sweepStart, sweepEnd
    });
    gen.generate();
    const output = args.o || args.output || 'output.wav';
    gen.saveWAV(output);
    console.log(`Saved to ${output}`);
    if (!args['no-play']) gen.play(output);
}

if (process.argv.length > 2) cli(); else interactive();
